src/test/spec/unified_runner/matcher.rs

```diff
@@ -0,0 +1,235 @@
+use crate::bson::{doc, spec::ElementType, Bson};
+
+pub fn results_match(actual: Option<&Bson>, expected: &Bson) -> bool {
+    match expected {
+        Bson::Document(expected_doc) => {
+            if let Some(special_op) = expected_doc.iter().next() {
+                if special_op.0.starts_with("$$") {
+                    return special_operator_matches(special_op, actual);
+                }
+            }
+            let actual = match actual {
+                Some(actual) => actual,
+                // The only case in which None is an acceptable value is if the expected document
+                // is a special operator; otherwise, the two documents do not match.
+                None => return false,
+            };
+            let actual_doc = actual.as_document().unwrap();
+            for (key, value) in expected_doc {
+                if !results_match(actual_doc.get(key), value) {
+                    return false;
+                }
+            }
+            true
+        }
+        Bson::Array(expected_array) => {
+            let actual_array = actual.unwrap().as_array().unwrap();
+            if expected_array.len() != actual_array.len() {
+                return false;
+            }
+
+            for (actual, expected) in actual_array.iter().zip(expected_array) {
+                if !results_match(Some(actual), expected) {
+                    return false;
+                }
+            }
+
+            true
+        }
+        _ => match actual {
+            Some(actual) => actual == expected,
+            None => false,
+        },
+    }
+}
+
+fn special_operator_matches(special_op: (&String, &Bson), actual: Option<&Bson>) -> bool {
```

Comment:
Instead of using `.0` and `.1`, you can destructure in the parameter list, e.g. `fn special_operator_matches((key, value): (&String, &Bson), ...)`